syntax = "proto3";

package geospatial;

// =============================================================================
// SOLICITUDES/RESPUESTAS DE ARCHIVOS (.CSV)
// =============================================================================

// Statistical information for a column
message ColumnStats {
    double count = 1;
    double mean = 2;
    double std = 3;
    double min = 4;
    double q25 = 5;    // 25th percentile
    double q50 = 6;    // 50th percentile (median)
    double q75 = 7;    // 75th percentile
    double max = 8;
    int32 null_count = 9;
    int32 unique_count = 10;
}

// Información de columna detectada automáticamente
message ColumnInfo {
    string name = 1;
    string type = 2; // "number" or "string"
    bool is_required = 3; // true for ID, X, Y, Z columns
    ColumnStats stats = 4; // Statistical information (for numeric columns)
    repeated string sample_values = 5; // Sample values from the column
}

// Solicitud para analizar las primeras filas y detectar columnas
message AnalyzeCsvRequest {
    string file_path = 1;
    string file_name = 2;
    int32 rows_to_analyze = 3; // default: 1000 for comprehensive statistics
    bool save_to_sql = 4; // optional: save analyzed data to SQL database
}

// Respuesta con información de columnas detectadas
message AnalyzeCsvResponse {
    repeated ColumnInfo columns = 1;
    map<string, string> auto_detected_mapping = 2; // suggested mappings like "id" -> "ID_COLUMN"
    bool success = 3;
    string error_message = 4;
    int32 total_rows = 5; // Total number of rows in the dataset
    int32 total_columns = 6; // Total number of columns
    double file_size_mb = 7; // File size in megabytes
    string encoding = 8; // File encoding detected
    repeated string numeric_columns = 9; // List of numeric column names
    repeated string categorical_columns = 10; // List of categorical column names
}

// Solicitud para procesar el archivo completo con mapping de variables
message SendFileRequest {
    string file_path = 1;
    string file_name = 2;
    string file_type = 3;
    string x_variable = 4;    // Column name for X coordinate
    string y_variable = 5;    // Column name for Y coordinate 
    string z_variable = 6;    // Column name for Z value (optional)
    string id_variable = 7;   // Column name for ID (optional)
    string depth_variable = 8; // Column name for DEPTH (optional)
    // Frontend preview overrides
    map<string, string> column_types = 9; // header -> "string" | "number"
    bool include_first_row = 10; // include first data row from preview in processing
    repeated string included_columns = 11; // columns selected in preview to process
}

// Respuesta después de procesar el archivo
message SendFileResponse {
    int32 total_rows_processed = 1;
    int32 valid_rows = 2;
    int32 invalid_rows = 3;
    repeated string errors = 4;
    bool success = 5;
    string processing_time = 6;
}

// =============================================================================
// CHUNKED ACCESS TO LOADED CSV DATA
// =============================================================================

message CsvDataRow {
    double x = 1;
    double y = 2;
    double z = 3;
    string id = 4;
    map<string, double> metrics = 5; // numeric properties (e.g., ppm_cu)
    map<string, string> attrs = 6;   // string properties
}

message GetLoadedDataChunkRequest {
    int32 offset = 1;
    int32 limit = 2;
}

message GetLoadedDataChunkResponse {
    repeated CsvDataRow rows = 1;
    int32 total_rows = 2;
    bool is_complete = 3;
    int32 next_offset = 4;
    repeated string available_metric_keys = 5;
}
